File: compiler/src/macros/error_macros.py
from dataclasses import replace
from core.macro_registry import MacroContext, Macro_emission_provider, Macro_typecheck_provider
from utils.strutil import IndentedStringIO

class Must_compile_error_macro_provider(Macro_emission_provider, Macro_typecheck_provider):
    def emission(self, ctx: MacroContext):
        """Process must_compile_error children during emission to catch emission-time errors.
        
        The children are processed normally but their output is discarded using dummy outputs.
        Any errors generated during emission will be available for the verification step to check.
        """
        # Create dummy outputs to discard emission results
        dummy_statement_out = IndentedStringIO()
        dummy_expression_out = IndentedStringIO()
        
        # Process all children with dummy outputs to catch potential emission-time errors
        for child in ctx.node.children:
            child_ctx = replace(ctx, node=child, statement_out=dummy_statement_out, expression_out=dummy_expression_out)
            ctx.current_step.process_node(child_ctx)

    def typecheck(self, ctx: MacroContext):
        # Process children to generate type errors during type checking step
        for child in ctx.node.children:
            child_ctx = replace(ctx, node=child)
            ctx.current_step.process_node(child_ctx)

File: compiler/src_ts/macros/error_macros.ts
// macros/error_macros.ts
import { Type_check_result } from "../compiler_types/proper_types.ts";
import {
  Macro_provider,
  REGISTER_MACRO_PROVIDERS,
  Register_macro_providers,
} from "../core/macro_registry.ts";
import { Emission_macro_context } from "../pipeline/steps/emission.ts";
import { Type_checking_context } from "../pipeline/steps/typechecking.ts";

export class Must_compile_error_macro_provider implements Macro_provider {
  [REGISTER_MACRO_PROVIDERS](via: Register_macro_providers): void {
    via(Type_checking_context, "must_compile_error", this.typecheck.bind(this));
    via(Emission_macro_context, "must_compile_error", this.emission.bind(this));
  }

  emission(ctx: Emission_macro_context): void {
    for (const child of ctx.node.children) {
      ctx.clone_with({
        node: child,
        statement_out: [],
        expression_out: [],
      }).apply();
    }
  }

  typecheck(ctx: Type_checking_context): Type_check_result {
    for (const child of ctx.node.children) {
      ctx.clone_with({ node: child }).apply();
    }
    return null;
  }
}
